mozfldp/model.py

```diff
@@ -0,0 +1,106 @@
+from sklearn.utils.multiclass import unique_labels
+
+import copy
+
+
+class SGDModel:
+    """Wrapper around `scikit-learn`'s SGD classifier allowing for external
+    updating of model weights and a modified training interface.
+
+    Other kwargs supplied to the constructor are passed to the underlying
+    classifier.
+    """
+
+    def __init__(self, **kwargs):
+        self.classifier = SGDClassifier(**kwargs)
+
+    def set_training_classes(self, all_training_labels):
+        """Specify the set of all known classes in the training set.
+
+        This is necessary in order to use `classifier.partial_fit()`. It should
+        be called at some point prior to starting model training.
+
+        all_training_labels: the labels in the training data (doesn't need to be
+            unique). This is needed to enable `partial_fit`.
+        """
+        self.classifier.classes_ = unique_labels(all_training_labels)
+
+    def get_clone(self, trained=False):
+        """Create a clone of this classifier.
+
+        trained: if `True`, maintains current state including trained model
+            weights, interation counter, etc. Otherwise, returns an unfitted
+            model with the same initialization params.
+
+        Returns a new `SGDModel` instance.
+        """
+        if trained:
+            new_classifier = copy.deepcopy(self.classifier)
+        else:
+            new_classifier = clone(self.classifier)
+            # If class labels have been specified, maintain these.
+            if getattr(self.classifier, "classes_", None):
+                new_classifier.classes_ = self.classifier.classes_
+
+        new_model = self.__class__()
+        new_model.classifier = new_classifier
+        return new_model
+
+    def __repr__(self):
+        return "SGDModel(\n{}\n)".format(self.classifier.__repr__())
+
+    def set_weights(self, coef, intercept):
+        """Update the current model weights.
+
+        Note that this leaves the iteration counters as-is. These are used
+        internally in computing an adaptive learning rate.
+        """
+        self.classifier.coef_ = coef
+        self.classifier.intercept_ = intercept
+
+    def get_weights(self):
+        """Return the current model weights as (coef, intercept)."""
+        return (self.classifier.coef_, self.classifier.intercept_)
+
+    def minibatch_update(self, X, y):
+        """Run a single weight update on the given minibatch.
+
+        X and y should be arrays of the appropriate dimensions as required by
+        `SGDClassifier.fit()`.
+        """
+        init_weight = [self.classifier.coef_, self.classifier.intercept_]
+        # array used to store sum of all the weights for current min-batch
+        mini_batch_coefs = np.zeros(init_weight[0].shape, dtype=np.float64, order="C")
+        mini_batch_intercept = np.zeros(init_weight[1].shape, dtype=np.float64, order="C")
+
+        for i in range(len(x)):
+            self.classifier.partial_fit(X[i],y[i])
```

Comment:
Subsetting`X` and `y` like this reduces the dimensionality, eg. `X[i]` will be a 1D array. We still want to maintain the original dimensionality. I think this can be achieved by using `X[i:(i+1)]`